File: visitor-pattern/python/src/ConcreteVisitorB.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""

from src.Visitor import Visitor


# 具体访问者B
class ConcreteVisitorB(Visitor):

    def __init__(self, ):
        Visitor.__init__(self)
        # 假如由不同厂商是程序的访问者
        self.name = 'Apple Visitor'

    def visit(self, concreteElement):
        print('ConcreteVisitorB::visit() [Element.class = ' +
              concreteElement.__class__.__name__ + ' Element.name = ' +
              concreteElement.get_name() + ']')
        concreteElement.operate()

    def get_name(self, ):
        return self.name

File: visitor-pattern/java/src/ConcreteVisitorB.java
package src;

// 具体访问者B
public class ConcreteVisitorB implements Visitor {

  // 假如由不同厂商是程序的访问者
  private String name = "Apple Visitor";

  @Override
  public void visit(ConcreteElementA concreteElementA) {
    System.out.println(
        "ConcreteVisitorB::visit() [Element.class = " + concreteElementA.getClass().getSimpleName()
            + " Element.name = "
            + concreteElementA.getName() + "]");
    concreteElementA.operate();
  }

  @Override
  public void visit(ConcreteElementB concreteElementB) {
    System.out.println(
        "ConcreteVisitorB::visit() [Element.class = " + concreteElementB.getClass().getSimpleName()
            + " Element.name = "
            + concreteElementB.getName() + "]");
    concreteElementB.operate();
  }

  public String getName() {
    return this.name;

  }
}